lib/auth/windows/objectsid.go

```diff
@@ -0,0 +1,96 @@
+// Modified from https://github.com/bwmarrin/go-objectsid
+// BSD 2-Clause License
+
+// Copyright (c) 2019, bwmarrin
+// All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+
+// 1. Redistributions of source code must retain the above copyright notice, this
+//    list of conditions and the following disclaimer.
+
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package windows
+
+import (
+	"fmt"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/gravitational/trace"
+)
+
+// adSID is an Active Directory ObjectSID
+type adSID struct {
+	RevisionLevel     int
+	SubAuthorityCount int
+	Authority         int
+	SubAuthorities    []int
+	RelativeID        *int
+}
+
+func (sid adSID) String() string {
+	s := fmt.Sprintf("S-%d-%d", sid.RevisionLevel, sid.Authority)
+	for _, v := range sid.SubAuthorities {
+		s += fmt.Sprintf("-%d", v)
+	}
+	return s
+}
+
+func decodeADSID(b []byte) (adSID, error) {
```

Comment:
Is there anything that you can link to that describes the format you're decoding here? Without a source it just looks like a bunch of arbitrary numbers.

Reply:
467e69d08364ffd741753130d3d9077dfab63cc1